### ISSUE TITLE
Parts of animated transform are silently ignored

### ISSUE THREAD

**Issue by gaearon** [software-mansion/react-native-reanimated#5116]
### Description

I've just spent a whole hour on this.

Reanimated offers two different APIs for passing styles:

- "Shortcut" `style={{ someProp: sharedValue }}` API
- `useAnimatedStyle`

I started with the shortcut API because [that's what the first guide teaches](https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/your-first-animation).
However, its limitations aren't clearly explained or warned about.

This is where I got stuck:

```js
  // Doesn't work
  const animatedScale = useSharedValue(0.3)
  return (
    <AnimatedImage style={{ transform: [{scale: animatedScale}] }} />
```

This code doesn't work because (I guess?) the "pass shared value directly" works only at the top level, i.e. for the entire rule. Apparently I can't use it for a part of the value (like `transform: [{ scale: ... }]`).

But there was no error or warning. The style was just silently being dropped. Of course, when I tried to console log it (to debug what's happening), I wasn't able to since I can't read `.value` on the JS thread. So that was doubly confusing because I had no tools to even begin understanding what is going wrong.

After a lot of back and forth I guessed that maybe I should `useAnimatedStyle` instead, and that worked.

```js
  // Works
  const animatedScale = useSharedValue(0.3)
  const animatedStyle = useAnimatedStyle(() => {
    return {
      transform: [{scale: animatedScale.value}],
    }
  })
  return (
    <AnimatedImage style={animatedStyle} />
```

What I'd like:

- Either it should be possible to nest shared values anywhere inside the `style` object.
- Or the library should clearly error and prevent me from passing them.

### Steps to reproduce

See above

### Snack or a link to a repository

sorry

### Reanimated version

3.4.2

### React Native version

0.72.4

### Platforms

Android

### JavaScript runtime

Hermes

### Workflow

None

### Architecture

None

### Build type

None

### Device

None

### Acknowledgements

Yes

**Comment by tjzel** [comment]
Hi @gaearon, thanks very much for your effort and interest in Reanimated!

Unfortunately, I tested it on both Android and iOS, on React Native 0.72.3 and .4, with Reanimated 3.4.2 and 3.5.4, using different components and not only Image - and it was working in all these configurations.


https://github.com/software-mansion/react-native-reanimated/assets/40713406/fa8efc81-7b84-4a4b-a627-8fac17d2921e

<details>
<summary>
Testing snippet
</summary>

```TSX
import React from 'react';
import {Button, StyleSheet, Image, View, Text} from 'react-native';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
} from 'react-native-reanimated';

const AnimatedImage = Animated.createAnimatedComponent(Image);

const doge = require('./doge.jpg');

export default function EmptyExample() {
  const animatedScale = useSharedValue(0.5);

  const animatedStyle = useAnimatedStyle(() => {
    return {
      transform: [{scale: animatedScale.value}],
    };
  });

  function handleOnPress() {
    animatedScale.value += 0.1;
  }
  return (
    <View style={styles.container}>
      <Text>Inline style</Text>
      <View style={styles.imagesContainer}>
        <Animated.Image
          style={[styles.image, {transform: [{scale: animatedScale}]}]}
          source={doge}
        />
        <AnimatedImage
          style={[styles.image, {transform: [{scale: animatedScale}]}]}
          source={doge}
        />
      </View>
      <Text>useAnimatedStyle</Text>
      <View style={styles.imagesContainer}>
        <Animated.Image style={[styles.image, animatedStyle]} source={doge} />
        <AnimatedImage style={[styles.image, animatedStyle]} source={doge} />
      </View>
      <Button title={'Press me'} onPress={handleOnPress} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: 200,
    alignItems: 'center',
  },
  imagesContainer: {
    flexDirection: 'row',
  },
  image: {
    height: 100,
    width: 100,
  },
});
```

</details>

**Comment by terrysahaidak** [comment]
Just tried to do the same with with regular view in Reanimated `3.5.2` and RN `0.72.3` - works just fine with scale transform.
What do you use as AnimatedImage?

**Comment by gaearon** [comment]
Thanks for checking. I was using Expo image but I think I also tried with vanilla RN image. Lemme try to reproduce again!

**Comment by sennevc** [comment]
I ran into the same issue as @gaearon and did some additional testing and managed to find a way to reproduce it.

When using the sharedValues with the "shortcut" API, a rerender will make them be silently ignored. A next rerender will apply them again. Afterwards there are no issues. This is not specific to the transform style props, but happens with all sharedValues.

Using the `useAnimatedStyle` hook seems to work, just like @gaearon pointed out in his original findings.

I managed to reproduce this issue on both android and iOS. In this demo the top box is using the shortcut API where as the bottom one uses the `useAnimatedStyle` hook:

https://github.com/software-mansion/react-native-reanimated/assets/25914162/70c90f42-3de2-4f26-b85b-df49abba8d39

<details>
<summary>Testing code</summary>

```typescript
import { StatusBar } from 'expo-status-bar';
import { Button, StyleSheet, View } from 'react-native';
import Animated, { useAnimatedStyle, useSharedValue, withSequence, withTiming } from 'react-native-reanimated'
import { useState } from 'react';

export default function App() {
  const [randomValue, setRandomValue] = useState(0)
  
  const boxOpacity = useSharedValue(0.5)
  const boxScale = useSharedValue(0.5)
  const boxAnimationStyle = useAnimatedStyle(() => ({
    opacity: boxOpacity.value,
    transform: [{scale: boxScale.value}]
  }))


  const startAnimation = () => {
    boxOpacity.value = withSequence(
      withTiming(1),
      withTiming(0),
      withTiming(0.5),
    )
    boxScale.value = withSequence(
      withTiming(1.5),
      withTiming(0.5),
      withTiming(1)
    )
  }

  const randomizeValue = () => {
    setRandomValue(Math.random())
  }

  console.log('Render')

  return (
    <View style={styles.container}>
      <View style={styles.boxContainer}>
        <Animated.View style={[styles.box, {opacity: boxOpacity, transform: [{scale: boxScale}]}]} />
        <Animated.View style={[styles.box, boxAnimationStyle]} />
      </View>
      <View style={styles.buttonContainer}>
        <Button title='force Rerender' onPress={randomizeValue} />
        <Button title='animate' onPress={startAnimation} />
      </View>
      <StatusBar style="auto" />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  boxContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-around',
  },
  box: {
    backgroundColor: 'red',
    width: 100,
    height: 100
  },
  buttonContainer: {
    flexDirection: 'row',
    padding: 20,
    justifyContent: 'space-evenly'
  }
});

```
</details>

I have `react-native-reanimated` version `3.3.0`

**Comment by tjzel** [comment]
I'm actually working on many improvements in Animated Components logic during renders in regards to Animated Props (so styles included) with or without inline usage and the fix for this issue will be there too.

**Comment by tjzel** [comment]
This was fixed with #5268
